### PR TITLE
Implemented balances monitoring API

### PR DESCRIPTION
**Pull request by DmitriyMolch**
https://cerenetwork.atlassian.net/browse/CBI-2093

### CODE REVIEW

File: api/src/lib/utils/index.js [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 
-function getTokenFloatAmount(amount, decimals){
-    const base = new BN(10);    
+function getTokenFloatAmount(amount, decimals) {
     const float = amount.divmod(base.pow(decimals));
     return parseFloat(`${float.div}.${float.mod}`);
-} 
+}
+
+function getCentsFromDecimals(amount, decimals) {
```
Comment (AndreiNavoichyk) [inline api/src/lib/utils/index.js]: Let's rename it to `toBaseUnits`.

File: api/src/lib/utils/index.js [hunk 2]
```diff
@@ -1,12 +1,17 @@
 const web3 = require('web3');
 const { BN } = web3.utils;
+const base = new BN(10);
 
-function getTokenFloatAmount(amount, decimals){
-    const base = new BN(10);    
+function getTokenFloatAmount(amount, decimals) {
```
Comment (AndreiNavoichyk) [inline api/src/lib/utils/index.js]: Let's rename to `toFloat`

File: api/src/services/healthService.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+const { blockchainNames, networkNames, accountGroups } = require('../config/blockchains');
+const accountsService = require('./accountsService');
+
+async function checkAccountsBalances(req, res) {  
+  try {
+    const { query } = req;
+    const blockchains = query.blockchains && query.blockchains.split('|');
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Any reasons to use `|` as a separator instead of `,`?

Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Use `url.parse` to get query params.

File: api/src/services/healthService.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    const networks = query.networks && query.networks.split('|');
+    const groups = query.groups && query.groups.split('|');
+    const addresses = query.addresses && query.addresses.split('|');
+    
+    if (!blockchains) {
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: 1. Move params checking to separate methods.
2. Do params getting one by one. It is no sense to extract addresses if there are no blockchains

File: api/src/services/healthService.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+          && (networks? networks.includes(account.network): true)
+          && (groups? groups.includes(account.group): true)
+          && (addresses? addresses.includes(account.address): true)
+          && account.balance.lt(account.minBalance)
+        ) {
+          belowMinAccounts.push(account);
+        }
+      });
+
+    if(belowMinAccounts.length) {
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Check formatting

File: api/src/services/healthService.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        msg: "Some accounts balances below the minimum",
+        accounts: belowMinAccounts
+      })
+    } else {
+      res.send({
+        msg: "No accounts balances below the minimum",
+      });
+    }
+  } catch (error) {
+    res.status(400).json({
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js:42]: Let's move this logic out of service.

File: api/src/services/healthService.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: The only reason is that `|` was in the API design

File: api/src/services/healthService.js (hunk 6 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js:42]: I did it in the same manner as it was done for `faucet` service. Do we need to change this structure later?

Comment (DmitriyMolch) [reply api/src/services/healthService.js:42]: if we want to separate controller and service then where should we do validation? In controller or service? Do we need to "invent the wheel" for this application?

File: api/src/services/healthService.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+          && (addresses? addresses.includes(account.address): true)
+          && account.balance.lt(account.minBalance)
+        ) {
+          belowMinAccounts.push(account);
+        }
+      });
+
+    if(belowMinAccounts.length) {
+      res.json({
+        msg: "Some accounts balances below the minimum",
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Let's replace message with more grammatically correct: `The following accounts have below the minimum balances`

File: api/src/services/healthService.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        accounts: belowMinAccounts
+      })
+    } else {
+      res.send({
+        msg: "No accounts balances below the minimum",
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Let's replace message `All accounts are good`

File: api/src/services/healthService.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: Why we need to use `url.parse` if params already parsed in the `query` variable?

File: api/src/services/healthService.js (hunk 6 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js:42]: If you take a look on `cereTokensService`, `metricsService` then see that it accepts `req`,  `res` objects. Other services don't have `req`, `res` objects.

File: api/src/lib/utils/index.js (hunk 1 above)
Comment (DmitriyMolch) [reply api/src/lib/utils/index.js]: Fixed

File: api/src/lib/utils/index.js (hunk 2 above)
Comment (DmitriyMolch) [reply api/src/lib/utils/index.js]: Fixed

File: api/src/services/healthService.js (hunk 4 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: Fixed

File: api/src/services/healthService.js (hunk 5 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

File: api/src/services/healthService.js (hunk 7 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

File: api/src/services/healthService.js (hunk 8 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

File: api/src/services/healthService.js (hunk 3 above)
Comment (AndreiNavoichyk) [reply api/src/services/healthService.js]: Let's use `,` and, please, update it in Notion.

File: api/src/services/healthService.js (hunk 6 above)
Comment (AndreiNavoichyk) [reply api/src/services/healthService.js:42]: Please, create a task to refactor services and get rid of res/req there.

File: api/src/services/healthService.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

File: api/src/services/healthService.js (hunk 6 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js:42]: Created  [a task](https://cerenetwork.atlassian.net/browse/CBI-2107)

File: api/src/services/healthService.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: The disadvantage of comma is that we can meet it often as a part of the text in query params. Let's imagine such URL `?search=hello, Men | hello, Girl`

Comment (Dmitriy1984Molchanov) [reply api/src/services/healthService.js:7]: fixed

File: api/src/services/accountsService.js [hunk 9]
```diff
@@ -1,7 +1,10 @@
-const { blockchainNames, tokenTypes } = require("../config/blockchains");
-const { blockchains } = require("../config");
+const Web3 = require('web3');
```
Comment (AndreiNavoichyk) [inline api/src/services/accountsService.js]: Why do you need `Web3` and `BN` here?

File: api/src/services/healthService.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    if (!supportedBlockchains.includes(blockchain)) {
+      throw new Error(`Incorrect "blockchains" param value "${blockchain}"`);
+    }
+  });
+}
+
+function validateNetworks(networks){
+  if (networks) {
+    const supportedNetworks = Object.values(networkNames);
+    networks.forEach(network =>{
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Fix formatting

File: api/src/services/healthService.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        throw new Error(`Incorrect "networks" param value "${network}"`);
+      }
+    });
+  }
+}
+
+function validateGroups(groups) {
+  if (groups) {
+    const supportedGroups = Object.values(accountGroups);
+    groups.forEach(group =>{
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Fix formatting

File: api/src/services/healthService.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      }
+    });
+  }
+}
+
+function validateAddresses(addresses, accounts) {
+  if (addresses) {
+    addresses.forEach(address => {
+      if (!accounts.find(account => account.address === address)) {
+        throw new Error(`Address "${address}" doesn't exist`);
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js:87]: Update the Notion with this message

File: api/src/services/healthService.js (hunk 3 above)
Comment (AndreiNavoichyk) [reply api/src/services/healthService.js]: In your example it is better to use query param encoding as far as `|` can be a part of the search as well.

File: api/src/services/healthService.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
+const { blockchainNames, networkNames, accountGroups } = require('../config/blockchains');
+const accountsService = require('./accountsService');
+
+async function checkAccountsBalances(req, res) {  
+  try {
+    const { query } = req;    
+    const blockchains = query.blockchains && query.blockchains.split(',');
```
Comment (AndreiNavoichyk) [inline api/src/services/healthService.js]: Move query params extracting to a separate method

File: api/src/services/healthService.js (hunk 3 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: Let's imagine query params are decoded to this string  `?search=hello, Men, hello, Girl`  and we use `,` as a separator. We will get 4 parts to search.

File: api/src/services/accountsService.js (hunk 9 above)
Comment (DmitriyMolch) [reply api/src/services/accountsService.js]: fixed

File: api/src/services/healthService.js (hunk 10 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

File: api/src/services/healthService.js (hunk 12 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js:87]: fixed

File: api/src/services/healthService.js (hunk 13 above)
Comment (DmitriyMolch) [reply api/src/services/healthService.js]: fixed

**Review by AndreiNavoichyk — APPROVED** [review]

**Pull request merged**